Objects/listobject.c

```diff
@@ -1970,6 +2167,88 @@ listsort_impl(PyListObject *self, PyObject *keyfunc, int reverse)
         lo.values = saved_ob_item;
     }
 
+
+    /* The pre-sort check: here's where we decide which compare function to use.
+     * How much optimization is safe? We test for homogeneity with respect to 
+     * several properties that are expensive to check at compare-time, and 
+     * set ms appropriately. */
+    if (saved_ob_size > 1) {
+        /* Assume the first element is representative of the whole list. */
+        int keys_are_in_tuples = (lo.keys[0]->ob_type == &PyTuple_Type &&
+                                  Py_SIZE(lo.keys[0]) > 0);
+
+        PyTypeObject* key_type = (keys_are_in_tuples ?
+                                  PyTuple_GET_ITEM(lo.keys[0],0)->ob_type :
+                                  lo.keys[0]->ob_type);
+
+        int keys_are_all_same_type = 1;
+        int strings_are_latin = 1;
+        int ints_are_bounded = 1;
+
+        /* Prove that assumption by checking every key. */
+        int i;
+        for (i=0; i< saved_ob_size; i++) {
+
+            if (keys_are_in_tuples &&
+                (lo.keys[i]->ob_type != &PyTuple_Type || Py_SIZE(lo.keys[i]) == 0)) {
+                keys_are_in_tuples = 0;
+                keys_are_all_same_type = 0;
+                break;
+            }
+
+            /* Note: for lists of tuples, key is the first element of the tuple
+             * lo.keys[i], not lo.keys[i] itself! We verify type-homogeneity 
+             * for lists of tuples in the if-statement directly above. */
+            PyObject* key = (keys_are_in_tuples ?
+                             PyTuple_GET_ITEM(lo.keys[i],0) :
+                             lo.keys[i]);
+
+            if (key->ob_type != key_type) {
+                keys_are_all_same_type = 0;
+                break;
+            }
+
+            else if (key_type == &PyLong_Type && ints_are_bounded &&
```

Comment:
`else` is not needed since an alternate branch is ended by `break`.

If `ints_are_bounded` or  `Py_ABS(Py_SIZE(key)) > 1` are false, `key_type == &PyUnicode_Type` still is checked.

I would write:
```
if (key_type == &PyLong_Type) {
    if (ints_are_bounded && Py_ABS(Py_SIZE(key)) > 1)
        ints_are_bounded = 0;
}
else if (key_type == &PyUnicode_Type ...
```